src/esn.calendar.libs/app/components/time-select/time-select.controller.js

```diff
@@ -0,0 +1,87 @@
+'use strict';
+
+angular.module('esn.calendar.libs')
+  .controller('calTimeSelectController', calTimeSelectController);
+
+function calTimeSelectController($scope, calMoment) {
+  const self = this;
+  let locale;
+  let timeFormat;
+
+  self.$onInit = onInit;
+  self.onSelectedTimeChange = onSelectedTimeChange;
+  self.onSetSelectedTime = onSetSelectedTime;
+  self.onTimeSelectBlur = onTimeSelectBlur;
+
+  function onInit() {
+    locale = self.locale;
+    timeFormat = self.timeFormat;
+    self.timeOptions = getTimeOptions();
+    self.isInputValid = true;
+    self.selectedTime = this.date.format(timeFormat);
+
+    // $watch for handling the offset being added to the date
+    $scope.$watch('ctrl.date', function(newDate) {
+      initTimeInput(newDate);
+    });
+  }
+
+  function getTimeOptions() {
+    const options = [];
+
+    [...Array(24).keys()].map(hour => {
+      options.push(calMoment({ hour }).locale(locale).format(timeFormat));
+      options.push(calMoment({ hour, minute: 15 }).locale(locale).format(timeFormat));
+      options.push(calMoment({ hour, minute: 30 }).locale(locale).format(timeFormat));
+      options.push(calMoment({ hour, minute: 45 }).locale(locale).format(timeFormat));
+    });
+
+    return options;
+  }
+
+  function initTimeInput(date) {
+    if (date) {
+      self.selectedTime = date.format(timeFormat);
+    }
+  }
+
+  function onSelectedTimeChange() {
+    if (calMoment(self.selectedTime, timeFormat, true).isValid()) {
+      const { hour, minute } = parseSelectedTime(self.selectedTime);
+
+      self.isInputValid = true;
+
+      self.date.set({
+        hour,
+        minute
+      });
+
+      self.onTimeChange();
+    } else {
+      self.isInputValid = false;
+    }
```

Comment:
This is how we often do it to avoid using too many indentations for a long block of code inside `if`:

```suggestion
    if (!calMoment(self.selectedTime, timeFormat, true).isValid()) {
      self.isInputValid = false;

      return;
    }

    const { hour, minute } = parseSelectedTime(self.selectedTime);

    self.isInputValid = true;

    self.date.set({
      hour,
      minute
    });

    self.onTimeChange();
```